File: Tree/509-pruneTree.cpp
#include<iostream>


class TreeNode
{
    public:
        int m_val;
        TreeNode* left{nullptr};
        TreeNode* right{nullptr};
        TreeNode(int v,TreeNode* l=nullptr,TreeNode* r=nullptr):m_val{v},left{l},right{r}{};
        void printTreeNodePreOrder(TreeNode* root)
        {
            if(!root)
                return;
            std::cout<<root->m_val<<'-';
            printTreeNodePreOrder(root->left);
            printTreeNodePreOrder(root->right);
        }
};


TreeNode* pruneTree(TreeNode* root)
{
    if(!root)
        return root;
    
    root->left=pruneTree(root->left);
    root->right=pruneTree(root->right);

    if(!root->left&&!root->right&&(root->m_val!=1))
        return nullptr;
    return root;
}


int main()
{
    TreeNode* root=new TreeNode(1);
    TreeNode* left=new TreeNode(0);
    TreeNode* right=new TreeNode(1);
    TreeNode* right2=new TreeNode(1);
    root->left=left;
    root->right=right;
    right->right=right2;
    root->printTreeNodePreOrder(root);
    TreeNode* res=pruneTree(root);
    std::cout<<'\n';
    res->printTreeNodePreOrder(res);
}

File: Tree/509-pruneTree.py
# Definition for a binary tree node.
'''
Given the root of a binary tree, return the same tree where every subtree (of the given tree) not containing a 1 has been removed.
A subtree of a node node is node plus every node that is a descendant of node.

Input: root = [1,null,0,0,1]
Output: [1,null,0,null,1]
Explanation: 
Only the red nodes satisfy the property "every subtree not containing a 1".
The diagram on the right represents the answer.

'''
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
    # def printTree(self):
    #     while 
class Solution(object):
    def pruneTree(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: Optional[TreeNode]
        """
        if not root:
            return
        
        root.left=self.pruneTree(root.left)
        root.right=self.pruneTree(root.right)
        
        if not root.left and not root.right and root.val!=1:
            return None
        return root
    

root=TreeNode(1)
left=TreeNode(0)
right=TreeNode(1)
right2=TreeNode(1)
root.left=left
root.right=right
right.right=right2

Solution().pruneTree(root)
